cloudsync/smartsync.py

```diff
@@ -0,0 +1,192 @@
+                 sleep: Optional[Tuple[float, float]] = None,
+                 root_oids: Optional[Tuple[str, str]] = None
+                 ):
+        super().__init__(providers=providers, roots=roots, storage=storage,
+                         sleep=sleep, root_oids=root_oids, smgr_class=SmartSyncManager, state_class=SmartSyncState)
+
+    def smart_unsync_ent(self, ent: SyncEntry):
+        # remove ent from requestset
+        # delete file, if it exists at path
+        self.state.smart_unsync_ent(ent)
+        if ent[LOCAL].sync_path:
+            ent_info = self.providers[LOCAL].info_path(ent[LOCAL].path)
+            if ent_info:
+                self.providers[LOCAL].delete(ent_info.oid)
+
+    def smart_unsync_path(self, path, side):
+        # TODO: should this take an oid? we may not have a local oid, so it might have to be a remote oid always...
+        path = self._make_path_remote(path, side)
+        ents = self.state.lookup_path(REMOTE, path)
+        if not ents:
+            raise FileNotFoundError(path)
+        for ent in ents:
+            self.smart_unsync_ent(ent)
+
+    def smart_sync_ent(self, ent):
+        """Request to sync down a file from the cloud, and mark the entry to maintain synchronization."""
+        # if the local file is missing,
+        #   clear the local side of the ent and
+        #   mark the remote side changed and clear the sync_path/sync_hash so that it will look new and sync down
+        #   this should override the local deletion event from before
+        # TODO: is there a race condition based on the local deletion event coming later? This might wrongly delete!
+        if ent[LOCAL].path and not self.providers[LOCAL].exists_path(ent[LOCAL].path):
+            ent[LOCAL].clear()
+            ent[REMOTE].sync_path = None
+            ent[REMOTE].sync_hash = None
+            self.state.update_entry(ent, REMOTE, None, changed=True)
+        self.state.smart_sync_ent(ent)
+        for parent_conflict in self.smgr.get_parent_conflicts(ent, REMOTE):  # ALWAYS remote
+            self.smgr.sync_one_entry(parent_conflict)
+        self.smgr.sync_one_entry(ent)
+
+    def smart_sync_path(self, path, side):
+        # TODO: should this take an oid? we don't have a local oid, so it would have to be a remote oid always...
+        path = self._make_path_remote(path, side)
+        ents = self.state.lookup_path(REMOTE, path)
+        if not ents:
+            raise FileNotFoundError(path)
+        for ent in ents:
+            self.smart_sync_ent(ent)
+
+    def smart_listdir_path(self, local_path):
+        # TODO: should this take an oid? we don't have a local oid, so it would have to be a remote oid always...
+        self.state: SmartSyncState
+        remote_path = self._make_path_remote(local_path, LOCAL)
+        listdir = self.state.smart_listdir(local_path, remote_path)
+        return listdir
+
+    def _make_path_remote(self, path, side):
+        if side == LOCAL:
```

Comment:
This seems strange to pass a side in here. I feel like the method name is explicit enough where it can just take a local path

Reply:
maybe it would be better called `_ensure_path_remote`. It checks the side here, rather than having to check the side every time the method is called

Reply:
Is the concern a caller could pass in a remote instead of a local path? Wouldn't that just fail to translate?

Reply:
It isn't a "concern", meaning it's not considered to be an error. Rather, the design is intended to be flexible allowing for passing either remote or local, and this method simply makes the path remote if it isn't already